### PR TITLE
enhancement(splunk_hec sink): Add configuration for source

### LINKED ISSUE #3251: Add configuration for HEC source field
In Splunk parlance, I think our `sourcetype` would typically be `_json`.

To allow users to set their own `sourcetype` for routing and parsing purposes I propose we:

* Add a `source` field to the `humio` and `splunk` sink types that can be use to set the source

The configuration would then look like:

```toml
[sinks.my_sink_id]
  # General
  type = "splunk_hec" # required
  inputs = ["my-source-or-transform-id"] # required
  healthcheck = true # optional, default
  host = "http://my-splunk-host.com" # required
  host_key = "hostname" # optional, no default
  indexed_fields = ["field1", "field2"] # optional, no default, relevant when encoding = "json"
  token = "${SPLUNK_HEC_TOKEN}" # required
  source = "/var/log/syslog" # optional, no default, will be automatically be set by the provider
```

By default, we will not send a `source` and allow the HEC provider set it.

### PR DESCRIPTION
**Pull request by jszwedko**
This adds a configuration option to set the `source` attribute of events
sent to Splunk. Typically this is the filename or network address where
the events came from.

https://docs.splunk.com/Splexicon:Source

Eventually, I could see us setting this automatically (maybe if they set
a sentinel value like `auto`) if the events come from the `file` source
(and possibly others as well), but this'll at least allow users to
configure it. They could set it to `{{ file }}` if the events came from
the `file` source to have it be the filename.

Fixes #3251

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Comment by binarylogic** [comment]
This is an instance where templating would be valuable, what do you think?

```
[sinks.splunk]
  type = "splunk"
  source = "{{file}}"
  # ...
```

**Comment by jszwedko** [comment]
@binarylogic Yeah, I think they could just template it as `{{ file }}` for now. I tried to note that in the PR description but it's kinda buried in there:

> They could set it to {{ file }} if the events came from the file source to have it be the filename.

I could still see an `auto` value in the future that would take a best guess based on the source of the events (to do things like `udp:514` for a syslog source).

**Comment by binarylogic** [comment]
Exactly, this is something that keeps popping up, the ability to automatically map data like that. Regarding templating, it doesn't appear that dynamic values are supported with your new option, correct? In other words, `{{file}}` would not work if a user supplied it.

**Comment by jszwedko** [comment]
> Regarding templating, it doesn't appear that dynamic values are supported with your new option, correct? In other words, {{file}} would not work if a user supplied it.

Ah, I see, for some reason I thought it was given for free, but I see that it needs to be wired up explicitly. I'll update.

**Comment by jszwedko** [comment]
I updated this to make the `source` field templatable. Do you mind taking a peek again @lukesteensen ?

File: src/internal_events/splunk_hec.rs [hunk 1]
```diff
@@ -44,3 +45,26 @@ impl InternalEvent for SplunkEventEncodeError {
         );
     }
 }
+
+#[derive(Debug)]
+pub struct SplunkSourceMissingKeys {
```
Comment (jszwedko) [inline src/internal_events/splunk_hec.rs:50]: I expect to open another PR making `sourcetype` configurable as well which I _think_ should be a separate event?

**Pull request merged**